### PR TITLE
Allow USB mode selection on USB connect

### PR DESCRIPTION
**Pull request by schwabe**

### CODE REVIEW

File: radio/src/targets/common/arm/stm32/CMakeLists.txt [hunk 1]
```diff
@@ -38,7 +36,6 @@ set(FIRMWARE_TARGET_SRC
   ../common/arm/stm32/usbd_usr.cpp
   ../common/arm/stm32/usb_driver.cpp
   )
-if(USB STREQUAL SERIAL)
   add_definitions(-DUSB_SERIAL)
```
Comment (projectkk2glider) [inline radio/src/targets/common/arm/stm32/CMakeLists.txt]: Indentation

File: radio/src/debug.h [hunk 2]
```diff
@@ -41,11 +41,9 @@ extern "C" {
 #elif defined(SEMIHOSTING)
   #include <stdio.h>
   #define debugPrintf(...) printf(__VA_ARGS__)
-#elif defined(DEBUG) && defined(CLI) && defined(USB_SERIAL)
-  #define debugPrintf(...) do { if (cliTracesEnabled) serialPrintf(__VA_ARGS__); } while(0)
 #elif defined(DEBUG) && defined(CLI)
-  uint8_t serial2TracesEnabled();
-  #define debugPrintf(...) do { if (serial2TracesEnabled() && cliTracesEnabled) serialPrintf(__VA_ARGS__); } while(0)
+   uint8_t serial2TracesEnabled();
+  #define debugPrintf(...) do { if (cliTracesEnabled) serialPrintf(__VA_ARGS__); if (serial2TracesEnabled() && cliTracesEnabled) serialPrintf(__VA_ARGS__); } while(0)
```
Comment (projectkk2glider) [inline radio/src/debug.h]: This looks wrong. The idea was/is:
 * in USB serial mode: call `serialPrintf` only if `cliTracesEnabled == true`
 * in real serial port mode: call `serialPrintf` if `(serial2TracesEnabled() && cliTracesEnabled)`

File: radio/src/gui/128x64/radio_setup.cpp [hunk 3]
```diff
@@ -563,6 +564,10 @@ void menuRadioSetup(event_t event)
         lcdDrawText(lcdLastRightPos, y, STR_MS, attr);
         if (attr) CHECK_INCDEC_GENVAR(event, g_eeGeneral.switchesDelay, -15, 100-15);
         break;
+
+      case ITEM_SETUP_USB_MODE:
+        g_eeGeneral.USBMode = editChoice(RADIO_SETUP_2ND_COLUMN, y, STR_USBMODE, STR_USBMODES, g_eeGeneral.USBMode, 0, 3, attr, event);
```
Comment (bsongis) [inline radio/src/gui/128x64/radio_setup.cpp]: perhaps an enum instead of 3 ?

File: radio/src/targets/common/arm/stm32/serial2_driver.cpp [hunk 4]
```diff
@@ -153,10 +153,11 @@ extern "C" void SERIAL_USART_IRQHandler(void)
...
-#if !defined(USB_SERIAL) && defined(CLI)
-  // Receive
-  uint32_t status = SERIAL_USART->SR;
-  while (status & (USART_FLAG_RXNE | USART_FLAG_ERRORS)) {
+#if defined(CLI)
+  if (!(getSelectedUsbMode() == USB_SERIAL_MODE)) {
+    // Receive
+    uint32_t status = SERIAL_USART->SR;
+    while (status & (USART_FLAG_RXNE | USART_FLAG_ERRORS)) {
     uint8_t data = SERIAL_USART->DR;
```
Comment (bsongis) [inline radio/src/targets/common/arm/stm32/serial2_driver.cpp]: indentation issue here

File: radio/src/targets/common/arm/stm32/usb_driver.h [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+#ifndef OPENTX_USB_DRIVER_H
+#define OPENTX_USB_DRIVER_H
+
+// USB driver
+enum usbMode {
+  USB_UNSELECTED_MODE,
+  USB_JOYSTICK_MODE,
+  USB_SERIAL_MODE,
+  USB_MASS_STORAGE_MODE,
```
Comment (bsongis) [inline radio/src/targets/common/arm/stm32/usb_driver.h:29]: here is a nice enum which needs to be used in the menus!

File: radio/src/targets/common/arm/stm32/usbd_desc.c [hunk 6]
```diff
@@ -28,28 +28,33 @@
 #include "usbd_core.h"
 #include "usbd_req.h"
 #include "usb_regs.h"
+#if defined(PCBTARANIS)
+#include "../../../taranis/board.h"
```
Comment (bsongis) [inline radio/src/targets/common/arm/stm32/usbd_desc.c]: isn't it possible to include board.h without the path here?

**Review by bsongis — APPROVED** [review]
Excellent PR!

**Comment by ReFil** [comment]
soooo how do we use this? i just installed @sirPerna 's built version of opentx for the baudrate switch, and it just goes to usb drive not simulator mode when i plug it in when powered on.

**Comment by schwabe** [comment]
@livid-fpv I have no idea how @sirPerna version was built. If you don't have the option to select usb mode in general settings, his build was done before this was merged.

Keep an eye on the downloads page we should release nightly builds soon.

**Comment by kilrah** [comment]
> soooo how do we use this?

You wait for the release of OpenTX 2.2.1.

**Comment by ReFil** [comment]
ok @schwabe  It was build on the 10th. Thanks for the information

**Pull request merged**